api/api-cache.php

```diff
@@ -0,0 +1,132 @@
+{
+    global $db;
+    $q = $db->query('SELECT * FROM api_cache WHERE endpoint = "' . $db->real_escape_string($endpoint) . '"');
+    if ($q) {
+        $r = $q->fetch_array(MYSQLI_ASSOC);
+        return $r;
+    }
+    return null;
+}
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint. Will not return if the entry is expired.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|false|null The first entry that matches, false if the entry is expired, or null if no entry exists
+ */
+function getCached(string $endpoint)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    if ($r) {
+        $config = $db->query('SELECT * FROM outgoing_request_cache_config WHERE id = ' . $r['config_id'])->fetch_array(MYSQLI_ASSOC);
+        $ttl = $config['ttl'];
+        $now = time();
+        $retrievalTime = strtotime($r['last_successfully_retrieved'] . 'UTC');
+        if ($now - $retrievalTime <= $ttl) {
+            return $r;
+        }
+        return false;
+    }
+    return null;
+}
+
+/**
+ * Adds or updates an entry into the cache table.
+ * @param string $endpoint The endpoint for the API request
+ * @param string $content The content of API request's response
+ * @param int|string $config The ID of the config or its name in the config table
+ */
+function insertCached(string $endpoint, string $content, $config)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    $queryString = "";
+    if ($r) {
+        $id = $r['id'];
+        $queryString = 'UPDATE api_cache 
+                        SET last_successfully_retrieved = UTC_TIMESTAMP, 
+                            last_attempted_retrieval = UTC_TIMESTAMP, 
+                            retry_count = 0, 
+                            content =\'' . $db->real_escape_string($content) . '\'
+                        WHERE id = ' . $id;
+
+    } else {
+        $configId = 1;
+        if (gettype($config) == 'integer') {
+            $configId = $config;
+        } else if (gettype($config) == 'string') {
+            $configId = $db->query('SELECT * FROM outgoing_request_cache_config WHERE config_name = "' . $db->real_escape_string($config) . '"')->fetch_array(MYSQLI_ASSOC)['id'];
+        }
+        $queryString = "INSERT INTO api_cache (endpoint, last_successfully_retrieved, last_attempted_retrieval, config_id, content) 
+                    values ('" . $db->real_escape_string($endpoint) . "', UTC_TIMESTAMP,
+                            UTC_TIMESTAMP," . $configId . ", '" . $db->real_escape_string($content) . "')";
+    }
+    $q = $db->query($queryString);
+    if ($q === false) {
+        error_log('Failed to update API cache: ' . $db->error);
+    }
+}
+
+/**
+ * Deletes an entry from the cache or deletes all entries from the cache.
+ * @param string|null $endpoint The endpoint of the API request to be removed. If null, empties the cache
+ */
+function clearCached(string $endpoint = null){
```

Comment:
I'd say leave this commented out if we don't use it, it's a bit risky to have a drop all in the code otherwise

Reply:
Changed in 7cb4d48